[workarround] Drivers from server

dpordomingo opened the pull request:
fix #89
blocks https://github.com/bblfsh/dashboard/pull/123

I followed the sugestions made by @mcuadros at https://github.com/bblfsh/dashboard/issues/89#issuecomment-351655057, so the backend will be able to call gRPC https://godoc.org/github.com/bblfsh/bblfshd/daemon/protocol to get the list of drivers

```
protocol.NewProtocolServiceClient(*grpc.ClientConn)
  .DriverStates(ctx, &protocol.DriverStatesRequest{})
```
at [fetchDrivers(ctx, protocol.ProtocolServiceClient) ([]*protocol.DriverImageState, error)](https://github.com/bblfsh/dashboard/pull/122/files?expand=1#diff-859e0853f2a781e75b1e126b60e71b9fR27)

### Deploy

To deploy the changes, it should be also changed the 
[src-d/charts/bblfsh-dashboard](https://github.com/src-d/charts/blob/master/bblfsh-dashboard), using the latest version of `bblfsh/bblfshd:v2.4.2` and its latest drivers: `python`, `java`, ` `javascript`, `ruby`, `go` and `php`.

To access the controll protocol from the backend, the `bblfsh/bblfshd` container should also expose the `9433` port, and be run with the extra parameters: `-ctl-network=tcp -ctl-address=0.0.0.0:9433`

The [procedure used to install the drivers (src-d/charts/bblfsh-dashboard/templates/install-bblfshd-drivers-configmap.yaml)](https://github.com/src-d/charts/blob/master/bblfsh-dashboard/templates/install-bblfshd-drivers-configmap.yaml#L11) must use the new control network:
```shell
bblfshctl driver install {{ $language }} {{ $driver.repository }}:{{ $driver.tag }}
    --ctl-endpoint=tcp --ctl-address=localhost:9433
```

### Testing it:
Run `bblfshd`:
```shell
docker run --privileged -d
    -p 9432:9432
    -p 9433:9433
    --name bblfsh
    bblfsh/bblfshd
    -ctl-network=tcp
    -ctl-address=0.0.0.0:9433
```
And the dashboard:
```shell
docker run
    -p 8080:80
    --link bblfsh
    bblfsh/dashboard
    -bblfsh-addr bblfsh:9432
    -bblfshctl-addr bblfsh:9433
```
The installed drivers must be listed adding the addr parameter to the `bblfshctl` command:
```shell
docker exec -it bblfsh
    bblfshctl driver list --ctl-endpoint=tcp --ctl-address=localhost:9433
```

bzz commented:
@smola 

>I guess this is a temporary fix until we get this functionality implemented in the clients, right?

exactly! https://github.com/bblfsh/sdk/issues/253

smacker commented on src/state/languages.js:
```diff
@@ -56,11 +58,14 @@ export const reducer = (state = initialState, action) => {
-  dispatch({ type: LOADING });
+export const load = () => (dispatch, getState) => {
+  const { options: { customServer, customServerUrl } } = getState();
+  const from = customServer ? customServerUrl : undefined;
```
we might need a selector for it, I remember we have such code in code base a few times already.

bzz commented:
Looks really good, 👍 @dpordomingo !

One question: is there a reason why it's now `POST /drivers` instead of `GET` ?

dpordomingo commented:
I did the same that done for `/version` (it sends the `bblfsh` server URL to the backend too)
The backend, on the other hand, is not using that URL because the one used for parse, and control, is not the same. (so it is part of the workarround, that will be solved by https://github.com/bblfsh/sdk/issues/253)

smacker commented:
codecov reports `-15.68%` code coverage. Do you think you can add few tests?

dpordomingo commented:
Yes @smacker I MUST add some tests to keep (at least) the original test coverage. I just wanted to gather some reviews before adding tests :dancing_men: 
I'll rebase and ping you once this is ready for a second pass.
Many thanks for your reviews !!!

dpordomingo commented:
Thanks for your review;
@smacker you're right about the benefit of having separated PR for:
- makefile refactor ( aabbafc + 2e4e223 )
- the changes required by the issue itself (the rest of the commit)

Feel free to confirm it if you prefer having separated PRs, and I'll do so.

Now the implementation is stable enough, I added some tests covering the new things; the coverage was slightly increased more than 2% https://codecov.io/gh/bblfsh/dashboard/pull/122/diff

dpordomingo commented:
Thanks for your review;
@smacker you're right about the benefit of having separated PR for:
- makefile refactor ( aabbafc + 2e4e223 )
- the changes required by the issue itself (the rest of the commit)

Feel free to confirm it if you prefer having separated PRs, and I'll do so.

New changes:
- rebased history considering comments in discussion above
- 2c89abe Now the implementation is stable enough, I added some tests covering the new things; the coverage was slightly increased more than 2% https://codecov.io/gh/bblfsh/dashboard/pull/122/diff
- 9112470 Add some docs about how to run and build...

rporres commented:
If I understand how this works, this will need changes in the way we deploy things if we want it work properly. Driver versions are currently pulled from the values in the helm deployment and are created inside the container. We can update them, but if the pod is recreated, the versions of the drivers will be those of the helm release and not whatever.

We will need attach a volume per pod and introduce logic to determine if we want to deploy drivers or not, which will definitely cause trouble.

I don't see a good way of this working. Please organise a meeting to discuss it.

dpordomingo commented:
Thanks @rporres for taking a look at this!
It is not required to update the drivers automatically, but only to deploy the new ones.
I'll open a PR against the bblfsh-dashboard charts with the expected changes and we'll be able to continue the discussion there.

mcuadros commented:
@rporres, why a volume? The drivers can be downloaded inside of the container. For me every deploy a new dashboard, bblfshd, requires downloading all the repositories.

We should consider, from now and forever, k8s as a stateless environment. As soon we accept this everything will go better.

mcuadros commented:
@dpordomingo also who takes care of the charts, is Infrastructure.

BTW I don't know why this topic is threated on this issue about a frontend thing, the server is the same and was working like this sinces months, this changes just reflects a functionally that was on CLI on the Web.

rporres commented:
@mcuadros I understood we wanted to update the driver versions from the dashboard. I was suggesting scenarios that could make that possible that if we needed it. I very much agree that k8s should be as stateless
